scrapy_zyte_smartproxy/middleware.py

```diff
@@ -52,13 +53,28 @@ def __init__(self, crawler):
         self.spider = None
         self._bans = defaultdict(int)
         self._saved_delays = defaultdict(lambda: None)
+        self._auth_url = None
 
     @classmethod
     def from_crawler(cls, crawler):
         o = cls(crawler)
         crawler.signals.connect(o.open_spider, signals.spider_opened)
         return o
 
+    def _make_auth_url(self, spider):
+        parsed_url = urlparse(self.url)
+        auth = self.get_proxyauth(spider)
+        if not auth.startswith(b'Basic '):
+            raise ValueError(
+                'Zyte Smart Proxy Manager only supports HTTP basic access '
+                'authentication, but %s.%s.get_proxyauth() returned %r'
+                % (self.__module__, self.__class__.__name__, auth)
+            )
+        user_and_colon = b64decode(auth[6:].strip()).decode('utf-8')
```

Comment:
Hi @Gallaecio
Looks good!
Actually, I am not familiar with the `scrapy-zyte-smartproxy` codebase, but noticed that `auth` is generated by [basic_auth_header](https://github.com/scrapy-plugins/scrapy-zyte-smartproxy/blob/2eb970904bdae492e3b9a536050af32d1f352b9b/scrapy_zyte_smartproxy/middleware.py#L175) which is using `base64.urlsafe_b64encode` and "ISO-8859-1" encoding - [link](https://github.com/scrapy/w3lib/blob/97ac45b075b83d50e78838391a91ee1b296475e9/w3lib/http.py#L104)

Maybe `base64.urlsafe_b64decode` and "ISO-8859-1" instead of "utf-8" would be safer here?
```suggestion
        user_and_colon = b64decode(auth[6:].strip()).decode('ISO-8859-1')
```

Reply:
Yes, that seems more future-proof.

Reply:
I am unsure about the encoding bit, though. Given UTF-8 matches ISO-8859-1 for ASCII characters, and UTF-8 is the most likely encoding to be used for non-ASCII characters, UTF-8 is probably safer than ISO-8859-1 here. And, in any case, I do not expect Smart Proxy Manager keys to start using non-ASCII characters any time soon.

Reply:
I’ve opened https://github.com/scrapy/w3lib/issues/181 because I believe `basic_auth_header` should use `b64decode`, but as far as this change is concerned that should have no effect, I think.

Reply:
I decided to go with https://github.com/scrapy-plugins/scrapy-zyte-smartproxy/pull/103/commits/aa244132559dfa2134550ca7fa363322c450650b so that, if e.g. someone implemented a subclass that fixes https://github.com/scrapy/w3lib/issues/181, this change will not suddenly break their code.

Reply:
I have thought about looking into proper [userinfo encoding](https://url.spec.whatwg.org/#userinfo-percent-encode-set), however, I think it may be best to leave that as is for now, and only revisit if the format of Smart Proxy Manager API keys ever changes.

To be honest, I hesitated about https://github.com/scrapy-plugins/scrapy-zyte-smartproxy/pull/103/commits/aab2b18b1d62bcf979fff951fd85d47328ff117b in the first place because with the current format of Smart Proxy Manager API keys it is a non-issue as well.

Reply:
yep, it is a non-issue with the current format of API keys